ampd/src/handlers/sui_verify_msg.rs

```diff
@@ -12,3 +29,322 @@ pub struct Message {
     pub source_address: SuiAddress,
     pub payload_hash: Hash,
 }
+
+#[derive(Deserialize, Debug)]
+#[try_from("wasm-messages_poll_started")]
+struct PollStartedEvent {
+    #[serde(rename = "_contract_address")]
+    contract_address: TMAddress,
+    poll_id: PollID,
+    source_gateway_address: SuiAddress,
+    messages: Vec<Message>,
+    participants: Vec<TMAddress>,
+}
+
+pub struct Handler<C, B>
+where
+    C: SuiClient + Send + Sync,
+    B: BroadcasterClient,
+{
+    worker: TMAddress,
+    voting_verifier: TMAddress,
+    rpc_client: C,
+    broadcast_client: B,
+}
+
+impl<C, B> Handler<C, B>
+where
+    C: SuiClient + Send + Sync,
+    B: BroadcasterClient,
+{
+    #[allow(dead_code)]
+    pub fn new(
+        worker: TMAddress,
+        voting_verifier: TMAddress,
+        rpc_client: C,
+        broadcast_client: B,
+    ) -> Self {
+        Self {
+            worker,
+            voting_verifier,
+            rpc_client,
+            broadcast_client,
+        }
+    }
+
+    async fn broadcast_votes(&self, poll_id: PollID, votes: Vec<bool>) -> Result<()> {
+        let msg = serde_json::to_vec(&ExecuteMsg::Vote { poll_id, votes })
+            .expect("vote msg should serialize");
+        let tx = MsgExecuteContract {
+            sender: self.worker.as_ref().clone(),
+            contract: self.voting_verifier.as_ref().clone(),
+            msg,
+            funds: vec![],
+        };
+
+        self.broadcast_client
+            .broadcast(tx)
+            .await
+            .change_context(Error::Broadcaster)
+    }
+}
+
+#[async_trait]
+impl<C, B> EventHandler for Handler<C, B>
+where
+    C: SuiClient + Send + Sync,
+    B: BroadcasterClient + Send + Sync,
+{
+    type Err = Error;
+
+    async fn handle(&self, event: &Event) -> Result<()> {
+        let PollStartedEvent {
+            contract_address,
+            poll_id,
+            source_gateway_address,
+            messages,
+            participants,
+            ..
+        } = match event.try_into() as error_stack::Result<_, _> {
+            Err(report) if matches!(report.current_context(), EventTypeMismatch(_)) => {
+                return Ok(());
+            }
+            event => event.change_context(Error::DeserializeEvent)?,
+        };
+
+        if self.voting_verifier != contract_address {
+            return Ok(());
+        }
+
+        if !participants.contains(&self.worker) {
+            return Ok(());
+        }
+
+        let deduplicated_tx_ids: HashSet<_> = messages.iter().map(|msg| msg.tx_id).collect();
+        let transaction_blocks = self
+            .rpc_client
+            .finalized_transaction_blocks(deduplicated_tx_ids)
+            .await
+            .change_context(Error::TxReceipts)?;
+
+        let votes = messages
```

Comment:
add a comment that explains that the handler doesn't make any assumptions about id duplication in the event, but the rpc call fails(?) if there are duplicates, therefore you have to do this convolution of deduplication and mapping